File: benchmarks/js/queen.js
// benchmark: queen (JS: runs under both jsc and node)
// Count solutions to 13-queens by bitmask backtracking (32-bit int ops).
"use strict";

function solve(cols, ld, rd, all) {
  if (cols === all) {
    return 1;
  }
  var count = 0;
  var poss = ~(cols | ld | rd) & all;
  while (poss !== 0) {
    var p = poss & (-poss);
    poss = poss - p;
    count += solve(cols | p, (ld | p) << 1, (rd | p) >> 1, all);
  }
  return count;
}

function workload() {
  var seed = [13];
  var bits = seed[0];
  var all = (1 << bits) - 1;
  return solve(0, 0, 0, all);
}

var emit = (typeof print === "function") ? print : console.log;
function nowMs() { return performance.now(); }
(function () {
  var WARMUP = 3, TIMED = 11, checksum = 0, i;
  for (i = 0; i < WARMUP; i++) { checksum = workload(); }
  var times = new Array(TIMED);
  for (i = 0; i < TIMED; i++) {
    var t0 = nowMs();
    checksum = workload();
    var t1 = nowMs();
    times[i] = t1 - t0;
  }
  times.sort(function (a, b) { return a - b; });
  var median = times[(TIMED - 1) >> 1];
  emit(String(checksum) + " " + (median / 1000).toFixed(9));
})();

File: benchmarks/lua/queen.lua
-- benchmark: queen (LuaJIT)
-- Count solutions to 13-queens by bitmask backtracking (bit library, 32-bit).

local bit = require("bit")
local band, bor, bnot, lshift, rshift = bit.band, bit.bor, bit.bnot, bit.lshift, bit.rshift

local function solve(cols, ld, rd, all)
  if cols == all then
    return 1
  end
  local count = 0
  local poss = band(bnot(bor(cols, bor(ld, rd))), all)
  while poss ~= 0 do
    local p = band(poss, -poss)
    poss = poss - p
    count = count + solve(bor(cols, p), lshift(bor(ld, p), 1), rshift(bor(rd, p), 1), all)
  end
  return count
end

local function workload()
  local seed = { [0] = 13 }
  local bits = seed[0]
  local all = lshift(1, bits) - 1
  return solve(0, 0, 0, all)
end

local WARMUP, TIMED = 3, 11
local checksum = 0
for _ = 1, WARMUP do checksum = workload() end
local times = {}
for i = 1, TIMED do
  local t0 = os.clock()
  checksum = workload()
  local t1 = os.clock()
  times[i] = t1 - t0
end
table.sort(times)
local median = times[math.floor((TIMED + 1) / 2)]
io.write(string.format("%d %.9f\n", checksum, median))
